docs: github webhook receivers (#4411)
akuity/kargo: docs/docs/50-user-guide/60-reference-docs/80-webhook-receivers/github/index.md

by verifying their signatures.

:::note
The following command can be used to generate a sufficiently secure secret:

```shell
openssl rand -base64 48 | tr -d '=+/' | head -c 32
```

:::

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: gh-wh-secret
  namespace: kargo-demo
stringData:
  secret: <your-secret-here>
---
apiVersion: kargo.akuity.io/v1alpha1
kind: ProjectConfig
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  webhookReceivers: 
    - name: gh-wh-receiver
      github:
        secretRef:
          name: gh-wh-secret
```

## Retrieving the Receiver's URL

Kargo will generate a hard-to-guess URL from the configuration. We can obtain 
this URL using the following command:

```shell
  kubectl \
    get projectconfigs \
    kargo-demo \
    -n kargo-demo \
    -o=jsonpath='{.status.webhookReceivers}'
```

## Registering with Github

When configuring on Github, you can configure either a webhook or an app. We will outline instructions for both, starting with webhooks.

### Webhooks

1. Navigating to the Webhooks Dashboard

    1. Navigate to <Hlt>Settings</Hlt>
    1. Click <Hlt>Webhooks</Hlt>

>     1. Click <Hlt>Add webhook</Hlt>

Reviewer: You're still not mentioning the starting point for this. You can address that and simultaneously reduce the number of steps with something like this:

```
1. Navigate to `https://github.com/<account>/<repository>/settings/hooks` where
   `<account>` has been replaced with your GitHub username or an organization
   for which you are an administrator and `repository` has been replaced with
   the name of a repository belonging to that account.
```

![Step 1](./img/webhooks/123.png "Settings")

2. Registering the Webhook

> 

Reviewer: I think we're missing the step to click "Add webhook".

select <Hlt>Packages</Hlt>. This requires that you have connected the repository
and package. For more information on connecting repositories and packages refer to the 
[Github Docs here](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
:::

![Step 5](./img/webhooks/5.png "Event Subscription")
  
Finally, confirm <Hlt>Active</Hlt> is toggled and click <Hlt>Add webhook</Hlt>.

![Step 6](./img/webhooks/6.png "Submit Form")

3. Verifying Connectivity

From your Webhooks Dashboard, click on the webhook you just created.

![Step 7](./img/webhooks/7.png "Created")

Navigate to <Hlt>Recent Deliveries</Hlt>.

![Step 8](./img/webhooks/8.png "Recent Deliveries")

Click on the <Hlt>ping</Hlt> event and ensure a successful response was returned.

![Step 9](./img/webhooks/9.png "Response")

### Apps

It may be tedious to configure webhooks for each of your Github repositories. 
You can instead opt to configure a 
[Github App](https://docs.github.com/en/apps); allowing you to receive events 
from all or select repositories.

> 1. Navigate to the Github Apps Dashboard
> 
>     1. Navigate to <Hlt>Settings</Hlt>. This will be listed in a dropdown menu that is toggled by clicking your Github avatar.
>     1. Click <Hlt>Developer Settings</Hlt>.This will be in the bottom left-hand corner of the settings dashboard.
>     1. Click <Hlt>Github Apps</Hlt>.

Reviewer: This can also be done for an org, which is probably the more likely thing for someone to be doing in a professional context. We should make this work for both.

Conveniently, all authenticated users can reach the relevant page for their own account using a single URL that works for everyone: `https://github.com/settings/apps`.

For orgs, it's `https://github.com/organizations/<org name>/settings/apps`, which is somewhat less convenient because of the substitution that needs to be made.

Ultimately, we might want to go with something along these lines:

```suggestion
1. To create a new GitHub App owned by your own account and installable into
   your own repositories, navigate to https://github.com/settings/apps.

    OR

    To create a new GitHub App owned by an organization, of which you are an
    administrator, and installable into that organization's repositories,
    navigate to `https://github.com/organizations/<org name>/settings/apps`,
    where `<org name>` has been replaced with the organization's actual name.
```

From there, I think all the steps are the same for either.

![Step 1](./img/apps/123.png "Settings")

> 2. Register a new Github App

Reviewer: I recommend using `1.` for every item in an ordered list. It spares us from having to re-number everything every time an item is added to the list or removed from it.

```suggestion
1. Register a new Github App
```

> 

Reviewer: I think we're missing a step to click "New GitHub App".

>     1. Set a unique name in the <Hlt>Github App name</Hlt> field.

Reviewer: It's worth mentioning that this must be _globally_ unique for all of GitHub.

    from the [Configuring the Receiver](#configuring-the-receiver) 
    step.
    ![Step 5](./img/apps/5.png "Configure Webhook")
    1. Scroll down to <Hlt>Configure Permissions</Hlt> -> 
    <Hlt>Repository Permissions</Hlt>
     :::note
      For the option to subscribe to repo push events, we will need 
      <Hlt>read + write</Hlt> access for the <Hlt>Contents</Hlt> permission. For the option to subscribe to registry push events(ghcr), we will need <Hlt>read + write</Hlt>
      access for the <Hlt>Packages</Hlt> permission.
    :::
    ![Step 7](./img/apps/7.png "Permissions - Contents")
    ![Step 8](./img/apps/8.png "Permissions - Packages")
    1. Scroll down to <Hlt>Subscribe to events</Hlt>
    1. Select options based on your choices for the previous step.
    :::note
      If you set the <Hlt>Contents</Hlt> permission to <Hlt>read + write</Hlt> 
      you will have the <Hlt>Push</Hlt> option available to you. If you set <Hlt>Packages</Hlt> to <Hlt>read + write</Hlt> you will have the <Hlt>Registry package</Hlt> option available to you.
    :::
    ![Step 9](./img/apps/9.png "Subscribe to Events")
    1. Finally, click <Hlt>Create GitHub App</Hlt>.

>     ![Step 10](./img/apps/10.png "Submit Form")

Reviewer: Both in terms of how it appears here in the source _and_ in how it renders, this all looks a bit cramped. Introducing additional line breaks between the items in the sub-lists will help combat the "wall of text" feeling.

---
sidebar_label: GitHub
---

# The GitHub Webhook Receiver

The following instructions will work for Github,

> Github Enterprise Cloud, and GitHub Enterprise Server.

Reviewer: We could probably be more detailed in this section.

It is likely worthwhile to introduce the receiver by describing the events it responds to and what it does in response to them. i.e. A push event refreshes any Warehouses subscribed to the repository from which the event originated. A package event refreshes any Warehouse subscribed to images from a ghcr.io package repository.

Aside: I've just realized the implementation is not yet covering the possibility of a package event coming from a chart repo. I will open a PR for that.

## Configuring the Receiver

> The Kargo GitHub webhook receiver will require a Kubernetes Secret. This Secret is required to contain a `secret` key in its data map. Github will sign payloads using this `secret` via HMAC signature. Our webhook receiver will use this `secret` to verify the signature.

Reviewer: ```suggestion
The GitHub webhook receiver will need to reference a Kubernetes `Secret` with a
`secret` key in its data map. This
[shared secret](https://en.wikipedia.org/wiki/Shared_secret) will be used by
GitHub to sign requests. The receiver will use it to authenticate those requests
by verifying their signatures.
```

:::note
The following command can be used to generate a sufficiently secure secret:

```shell

> echo "$(openssl rand -base64 48 | tr -d '=+/' | head -c 32)"

Reviewer: `echo $(...)` is superfluous here. The command has the same effect without it.

```suggestion
openssl rand -base64 48 | tr -d '=+/' | head -c 32
```

```

:::

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: gh-wh-secret
  namespace: kargo-demo
stringData:
  secret: <your-secret-here>
---
apiVersion: kargo.akuity.io/v1alpha1
kind: ProjectConfig
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  webhookReceivers: 
    - name: gh-wh-receiver
      github:
        secretRef:
          name: gh-wh-secret
```

> ## Retrieving the Webhook URL

Reviewer: ```suggestion
## Retrieving the Receiver's URL
```

---
sidebar_label: GitHub
---

> # GitHub

Reviewer: ```suggestion
# The GitHub Webhook Receiver
```

The following instructions will work for Github,
Github Enterprise Cloud, and GitHub Enterprise Server.

> ## GitHub Webhook Receiver Configuration

Reviewer: ```suggestion
## Configuring the Receiver
```

The Kargo GitHub webhook receiver will require a Kubernetes Secret. This Secret is required to contain a `secret` key in its data map. You will be required to provide the value assigned to the `secret` key to GitHub in a later step so keep it handy.

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: gh-wh-secret
  namespace: kargo-demo
stringData:
  secret: your-secret-here
---
apiVersion: kargo.akuity.io/v1alpha1
kind: ProjectConfig
metadata:

>   name: my-project-config

Reviewer: ProjectConfig resources are required to be named the same as the Project and namespace. (This requirement naturally enforces a maximum of one ProjectConfig per Project.)

```suggestion
  name: kargo-demo
```

  name: gh-wh-secret
  namespace: kargo-demo
stringData:

>   secret: your-secret-here

Reviewer: This would provide a stronger visual cue that there's a blank to fill in here. Even if someone blindly copies it, they'll easily discover their mistake when troubleshooting.

```suggestion
  secret: <your secret here>
```

      github:
        secretRef:
          name: gh-wh-secret
```

Github will sign payloads using this `secret` via HMAC signature. Our webhook receiver will use this secret to verify the signature.

:::note
The following command can be used to generate a sufficiently secure secret:

```shell
echo "$(openssl rand -base64 48 | tr -d '=+/' | head -c 32)"
```

> :::

Reviewer: It feels awkward to explain what this is and how to generate _after_ the YAML into which this would be inserted.

If we were to lead off with the explanation of how the secret is used and the suggested command for generating it, then the reader has it already in hand by the time they reach the YAML.

> Github will sign payloads using this `secret` via HMAC signature. Our webhook receiver will use this secret to verify the signature.

Reviewer: I would still move this higher. Tell users what the secret's for exactly before showing them how to generate.

## Retrieving the Webhook URL

The secret (among other things) will be used as an input to generate
a secure URL for our newly created webhook receiver. We can obtain

> this URL using the following command:

Reviewer: This might be too much detail. idk if we are going to want to repeat this over and over again for each type of receiver. I might suggest just saying Kargo will generate a hard-to-guess URL ("secure" sounds like too much of a promise) from the configuration and here's how it can be retrieved...

You could link users to `../../20-how-to-guides/30-working-with-warehouses.md#configuring-a-receiver` in an info box if they want more details.

  namespace: kargo-demo
stringData:
  secret: <your-secret-here>
---
apiVersion: kargo.akuity.io/v1alpha1
kind: ProjectConfig
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  webhookReceivers: 
    - name: gh-wh-receiver
      github:
        secretRef:
          name: gh-wh-secret
```

## Retrieving the Receiver's URL

Kargo will generate a hard-to-guess URL from the configuration. We can obtain 
this URL using the following command:

```shell
  kubectl \
    get projectconfigs \
    kargo-demo \
    -n kargo-demo \
    -o=jsonpath='{.status.webhookReceivers}'
```

## Registering with Github

When configuring on Github, you can configure either a webhook or an app. We will outline instructions for both, starting with webhooks.

### Webhooks

1. Navigate to `https://github.com/<account>/<repository>/settings/hooks` where
   `<account>` has been replaced with your GitHub username or an organization
   for which you are an administrator and `repository` has been replaced with
   the name of a repository belonging to that account.

1. Click <Hlt>Add webhook</Hlt>.

![Step 1](./img/webhooks/123.png "Settings")

1. Complete the <Hlt>Add webhook</Hlt> form:

>    1. Set <Hlt>Payload URL</Hlt> to the URL we

Reviewer: ```suggestion

    ![Add webook](./img/webhooks/4.png)

    1. Set <Hlt>Payload URL</Hlt> to the URL we
       [retrieved earlier](#retrieving-the-receivers-url).

    1. Set <Hlt>Content type</Hlt> to `application/json`.

    1. Set <Hlt>Secret</Hlt> to the value previously assigned to the `secret`
       key of the `Secret` referenced by the
       [webhook receiver's configuration](#configuring-the-receiver).

    1. Under <Hlt>Which events would you like to trigger this webhook?</Hlt>:

        Leave <Hlt>Just the push event.</Hlt> selected, unless you would
        like to receive events when container images or Helm charts are
        pushed to associated GHCR repositories.

        To receive such events, select
        <Hlt>Let me select individual events.</Hlt>, then ensure
        <Hlt>Pushes</Hlt> and <Hlt>Packages</Hlt> are both checked.

        :::note
        You will only receive events from those GHCR repositories explicitly
        associated with your Git repository.

        For more information on this topic, refer to
        [these GitHub docs](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
        :::

    1. Ensure <Hlt>Active</Hlt> remains checked.

    1. Click <Hlt>Add webhook</Hlt>.
```

       [retrieved earlier](#retrieving-the-receivers-url).
    1. Set <Hlt>Content type</Hlt> to `application/json`.
    1. Set <Hlt>Secret</Hlt> to the value previously assigned to the `secret`
       key of the `Secret` referenced by the
       [webhook receiver's configuration](#configuring-the-receiver).

![Step 3](./img/webhooks/4.png "Add Webhook")

:::note
Leave the <Hlt>Just the push event</Hlt> field checked unless you're
looking to subscribe to `ghcr` events. If you're looking to subscribe to `ghcr` events you should select <Hlt>Let me select individual events</Hlt> and then 
select <Hlt>Packages</Hlt>. This requires that you have connected the repository
and package. For more information on connecting repositories and packages refer to the 
[Github Docs here](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
:::

![Step 5](./img/webhooks/5.png "Event Subscription")
  
Finally, confirm <Hlt>Active</Hlt> is toggled and click <Hlt>Add webhook</Hlt>.

> ![Step 6](./img/webhooks/6.png "Submit Form")

Reviewer: ```suggestion

    ![Add webook](./img/webhooks/4.png)

    1. Set <Hlt>Payload URL</Hlt> to the URL we
       [retrieved earlier](#retrieving-the-receivers-url).

    1. Set <Hlt>Content type</Hlt> to `application/json`.

    1. Set <Hlt>Secret</Hlt> to the value previously assigned to the `secret`
       key of the `Secret` referenced by the
       [webhook receiver's configuration](#configuring-the-receiver).

    1. Under <Hlt>Which events would you like to trigger this webhook?</Hlt>:

        Leave <Hlt>Just the push event.</Hlt> selected, unless you would
        like to receive events when container images or Helm charts are
        pushed to associated GHCR repositories.

        To receive such events, select
        <Hlt>Let me select individual events.</Hlt>, then ensure
        <Hlt>Pushes</Hlt> and <Hlt>Packages</Hlt> are both checked.

        :::note
        You will only receive events from those GHCR repositories explicitly
        associated with your Git repository.

        For more information on this topic, refer to
        [these GitHub docs](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
        :::

    1. Ensure <Hlt>Active</Hlt> remains checked.

    1. Click <Hlt>Add webhook</Hlt>.
```

    1. Click <Hlt>Add Webhook</Hlt>

![Step 1](./img/webhooks/123.png "Settings")

2. Registering the Webhook

    1. Set the <Hlt>Payload URL</Hlt> to the value we retrieved from the 
    [Retrieving the Receiver's URL](#retrieving-the-receivers-url) section.

    1. Set the <Hlt>Content type</Hlt> field to `application/json`.

    1. Set the <Hlt>Secret</Hlt> field to the value we assigned the `secret` key 
    from the [Configuring the Receiver](#configuring-the-receiver) section.

![Step 3](./img/webhooks/4.png "Add Webhook")

:::note
Leave the <Hlt>Just the push event</Hlt> field checked unless you're
looking to subscribe to `ghcr` events. If you're looking to subscribe to `ghcr` events you should select <Hlt>Let me select individual events</Hlt> and then 
select <Hlt>Packages</Hlt>. This requires that you have connected the repository
and package. For more information on connecting repositories and packages refer to the 
[Github Docs here](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
:::

![Step 5](./img/webhooks/5.png "Event Subscription")
  
Finally, confirm <Hlt>Active</Hlt> is toggled and click <Hlt>Add webhook</Hlt>.

> ![Step 6](./img/webhooks/6.png "Submit Form")

Reviewer: ```suggestion
    1. Set <Hlt>Payload URL</Hlt> to the URL we
       [retrieved earlier](#retrieving-the-receivers-url).

    1. Set <Hlt>Content type</Hlt> to `application/json`.

    1. Set <Hlt>Secret</Hlt> to the value previously assigned to the `secret`
       key of the `Secret` referenced by the
       [webhook receiver's configuration](#configuring-the-receiver).

    1. Under <Hlt>Which events would you like to trigger this webhook?</Hlt>:

        Leave <Hlt>Just the push event.</Hlt> selected, unless you would
        like to receive events when container images or Helm charts are
        pushed to associated GHCR repositories.

        To receive such events, select
        <Hlt>Let me select individual events.</Hlt>, then ensure
        <Hlt>Pushes</Hlt> and <Hlt>Packages</Hlt> are both checked.

        :::note
        You will only receive events from those GHCR repositories explicitly
        associated with your Git repository.

        For more information on this topic, refer to
        [these GitHub docs](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
        :::

    1. Ensure <Hlt>Active</Hlt> remains checked.

    1. Click <Hlt>Add webhook</Hlt>.
```

---
sidebar_label: GitHub
---

# The GitHub Webhook Receiver

> The Github Webhook Receiver will respond to ping, push, and package events.
> For ping events, a 200 is returned to verify connectivity. Push events refresh
> any Warehouses subscribed to the repository from which the event originated. A
> package event refreshes any Warehouses subscribed to images from a ghcr.io
> package registry.

Reviewer: ```suggestion
--------------------------------------------------------------------------------
The Github Webhook Receiver will respond to `ping`, `push`, and `package`
originating from GitHub repositories. 

The receiver unconditionally responds to `ping` events with an HTTP `200` status
code.

In response a `push` event, the receiver "refreshes" `Warehouse`s subscribed
to the Git repository from which the event originated.

In response to a `package` event, the receiver "refreshes" `Warehouse`s
subscribed to the GHCR repository from which the event originated.

:::info
"Refreshing" a `Warehouse` resource means enqueuing it for immediate
reconciliation by the Kargo controller, which will execute the discovery of
new artifacts from all repositories to which that `Warehouse` subscribes.
:::
```

Enterprise Server.
:::

## Configuring the Receiver

The GitHub webhook receiver will need to reference a Kubernetes `Secret` with a
`secret` key in its data map. This
[shared secret](https://en.wikipedia.org/wiki/Shared_secret) will be used by
GitHub to sign requests. The receiver will use it to authenticate those requests
by verifying their signatures.

:::note
The following command is suggested for generating a complex secret:

```shell
openssl rand -base64 48 | tr -d '=+/' | head -c 32
```

:::

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: gh-wh-secret
  namespace: kargo-demo
stringData:
  secret: <your-secret-here>
---
apiVersion: kargo.akuity.io/v1alpha1
kind: ProjectConfig
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  webhookReceivers: 
    - name: gh-wh-receiver
      github:
        secretRef:
          name: gh-wh-secret
```

## Retrieving the Receiver's URL

Kargo will generate a hard-to-guess URL from the configuration. We can obtain 
this URL using the following command:

```shell
  kubectl \
    get projectconfigs \
    kargo-demo \
    -n kargo-demo \
    -o=jsonpath='{.status.webhookReceivers}'
```

## Registering with Github

> When configuring on Github, you can configure either a webhook or an app. We will outline instructions for both, starting with webhooks.

Reviewer: ```suggestion
There are two options whereby GitHub repositories can be configured to send
events to the webhook receiver:

    1. Configure webhooks directly for a single repository.

        The advantage of this approach is that is is comparatively simple,
        however, its large disadvantage is that it is tedious and most
        likely infeasible to repeat this configuration for a large number
        of repositories.

    1. Create a [Github App](https://docs.github.com/en/apps).

        The disadvantage of this approach is that it is comparatively
        complex, however, its large advantage is that once created and
        configured, the App can be easily [installed]() into any number of
        GitHub repositories (belonging to the same account that owns the App).

In the following sections, we will present instructions for both options.
```

### Webhooks

> 

Reviewer: ```suggestion

To configure a single repository to notify the receiver of relevant events:

```

1. Navigate to `https://github.com/<account>/<repository>/settings/hooks`, where
   `<account>` has been replaced with your GitHub username or an organization
   for which you are an administrator and `repository` has been replaced with
   the name of a repository belonging to that account.

1. Click <Hlt>Add webhook</Hlt>.

> ![Step 1](./img/webhooks/123.png "Settings")

Reviewer: ```suggestion
    ![Step 1](./img/webhooks/123.png "Settings")
```

1. Complete the <Hlt>Add webhook</Hlt> form:

> 

Reviewer: ```suggestion

    ![Add webook](./img/webhooks/4.png)

```

    1. Set <Hlt>Payload URL</Hlt> to the URL we
        [retrieved earlier](#retrieving-the-receivers-url).

    1. Set <Hlt>Content type</Hlt> to `application/json`.

    1. Set <Hlt>Secret</Hlt> to the value previously assigned to the `secret`
        key of the `Secret` referenced by the
        [webhook receiver's configuration](#configuring-the-receiver).

    1. Under <Hlt>Which events would you like to trigger this webhook?</Hlt>:
        Leave <Hlt>Just the push event.</Hlt> selected, unless you would
        like to receive events when container images or Helm charts are
        pushed to associated GHCR repositories.

>         To receive such events, select

Reviewer: ```suggestion

        To receive such events, select
```

        pushed to associated GHCR repositories.

        To receive such events, select
        <Hlt>Let me select individual events.</Hlt>, then ensure
        <Hlt>Pushes</Hlt> and <Hlt>Packages</Hlt> are both checked.
        :::note

        You will only receive events from those GHCR repositories explicitly
        associated with your Git repository.
        For more information on this topic, refer to
        [these GitHub docs](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
        :::

    1. Ensure <Hlt>Active</Hlt> remains checked.

    1. Click <Hlt>Add webhook</Hlt>.

1. Verify connectivity:

    1. From the <Hlt>Webhooks</Hlt> dashboard, select the webhook you registered
       above.

>         ![Step 7](./img/webhooks/7.png "Created")

Reviewer: ```suggestion

        ![Step 7](./img/webhooks/7.png "Created")
```

>     1. Select the <Hlt>Recent Deliveries</Hlt> tab.

Reviewer: ```suggestion

    1. Select the <Hlt>Recent Deliveries</Hlt> tab.
```

        ![Step 8](./img/webhooks/8.png "Recent Deliveries")

    1. Select the <Hlt>ping</Hlt> event and ensure an HTTP response status of
       `200` was received from the webhook receiver.

>         ![Step 9](./img/webhooks/9.png "Response")

Reviewer: ```suggestion

        ![Step 9](./img/webhooks/9.png "Response")
```

> 

Reviewer: ```suggestion
```

> 

Reviewer: ```suggestion
When these steps are complete, the repository will send events to the webhook
receiver.

```

    from the [Configuring the Receiver](#configuring-the-receiver) 
    step.
    
    1. Scroll down to <Hlt>Configure Permissions</Hlt> -> 
    <Hlt>Repository Permissions</Hlt>.

     :::note
      For the option to subscribe to repo push events, we will need 
      <Hlt>read + write</Hlt> access for the <Hlt>Contents</Hlt> permission. For the option to subscribe to registry push events(ghcr), we will need <Hlt>read + write</Hlt>
      access for the <Hlt>Packages</Hlt> permission.
    :::

    ![Step 7](./img/apps/7.png "Permissions - Contents")

    ![Step 8](./img/apps/8.png "Permissions - Packages")

    1. Scroll down to <Hlt>Subscribe to events</Hlt>

    1. Select options based on your choices for the previous section.

    :::note
      If you set the <Hlt>Contents</Hlt> permission to <Hlt>read + write</Hlt> 
      you will have the <Hlt>Push</Hlt> option available to you. If you set <Hlt>Packages</Hlt> to <Hlt>read + write</Hlt> you will have the <Hlt>Registry package</Hlt> option available to you.
    :::

    ![Step 9](./img/apps/9.png "Subscribe to Events")

    1. Finally, click <Hlt>Create GitHub App</Hlt>.
    ![Step 10](./img/apps/10.png "Submit Form")

> 1. Verifying Connectivity
>   
>     1. In your new App's dashboard, click <Hlt>Advanced</Hlt> on the left-hand 
>     side.
> 
>     1. Click <Hlt>Recent Deliveries</Hlt>.
> 
>     1. Click on the <Hlt>ping</Hlt> event.
> 
>     ![Step 11](./img/apps/11.png "Recent Deliveries")
> 
>     1. Click <Hlt>Response</Hlt> and verify a 200 was received.
> 
>     ![Step 12](./img/apps/12.png "Response")

Reviewer: ```suggestion
1. Verify connectivity:
    1. From the new App's settings page, select <Hlt>Advanced</Hlt> from the
       left sidebar.

    1. Under <Hlt>Recent Deliveries</Hlt>, select the <Hlt>ping</Hlt> event and
       ensure an HTTP response status of `200` was received from the webhook
       receiver.

        ![Step 12](./img/apps/12.png "Response")
```

> #### Additional Documentation
> 
> For more additional information on configuring Github Webhooks or Apps, refer 
> to the [Github Docs](https://docs.github.com/en/webhooks/using-webhooks/creating-webhooks)
> 

Reviewer: Moved this up higher.

```suggestion
#### Additional Documentation

For more additional information on configuring Github Webhooks or Apps, refer 
to the [Github Docs](https://docs.github.com/en/webhooks/using-webhooks/creating-webhooks)

```

From your Webhooks Dashboard, click on the webhook you just created.

![Step 7](./img/webhooks/7.png "Created")

Navigate to <Hlt>Recent Deliveries</Hlt>.

![Step 8](./img/webhooks/8.png "Recent Deliveries")

Click on the <Hlt>ping</Hlt> event and ensure a successful response was returned.

![Step 9](./img/webhooks/9.png "Response")

### Apps

It may be tedious to configure webhooks for each of your Github repositories. 
You can instead opt to configure a 
[Github App](https://docs.github.com/en/apps); allowing you to receive events 
from all or select repositories.

1. Navigate to Github Apps dashboard

    1. To create a new GitHub App owned by your own account and installable into
   your own repositories, navigate to https://github.com/settings/apps.
    OR
    To create a new GitHub App owned by an organization, of which you are an
    administrator, and installable into that organization's repositories,
    navigate to `https://github.com/organizations/<org name>/settings/apps`,
    where `<org name>` has been replaced with the organization's actual name.
    1. From there, click <Hlt>New Github App</Hlt>.

![Step 1](./img/apps/123.png "Settings")

> 1. Register a new Github App
> 
>     ![Step 4](./img/apps/4.png "Register New App")
> 
>     1. Set a unique name in the <Hlt>Github App name</Hlt> field.
> 
>     :::note
>     This must be globally unique for all of GitHub.
>     :::
> 
>     1. For the <Hlt>Homepage URL</Hlt> field you can use an arbitrary URL
>     or the repository URL.
> 
>     ![Step 5](./img/apps/5.png "Configure Webhook")
> 
>       1. Set the <Hlt>Webhook URL</Hlt> to the value we retrieved from the 
>     [Retrieving the Receiver's URL](#retrieving-the-receivers-url) section.
> 
>     1. Set the <Hlt>Secret</Hlt> field to the value we set to the `secret` key 
>     from the [Configuring the Receiver](#configuring-the-receiver) 
>     step.
>     
>     1. Scroll down to <Hlt>Configure Permissions</Hlt> -> 
>     <Hlt>Repository Permissions</Hlt>.
> 
>      :::note
>       For the option to subscribe to repo push events, we will need 
>       <Hlt>read + write</Hlt> access for the <Hlt>Contents</Hlt> permission. For the option to subscribe to registry push events(ghcr), we will need <Hlt>read + write</Hlt>
>       access for the <Hlt>Packages</Hlt> permission.
>     :::
> 
>     ![Step 7](./img/apps/7.png "Permissions - Contents")
> 
>     ![Step 8](./img/apps/8.png "Permissions - Packages")
> 
>     1. Scroll down to <Hlt>Subscribe to events</Hlt>
> 
>     1. Select options based on your choices for the previous section.
> 
>     :::note
>       If you set the <Hlt>Contents</Hlt> permission to <Hlt>read + write</Hlt> 
>       you will have the <Hlt>Push</Hlt> option available to you. If you set <Hlt>Packages</Hlt> to <Hlt>read + write</Hlt> you will have the <Hlt>Registry package</Hlt> option available to you.
>     :::
> 
>     ![Step 9](./img/apps/9.png "Subscribe to Events")
> 
>     1. Finally, click <Hlt>Create GitHub App</Hlt>.
>     ![Step 10](./img/apps/10.png "Submit Form")

Reviewer: ```suggestion
1. Complete the <Hlt>Register new GitHub App</Hlt> form:

    1. Set <Hlt>Github App name</Hlt> to a name of your choosing.

        :::note
        This name must be _globally unique_ across all of GitHub.

        Unfortunately, you will not learn whether your selected name is
        unavailable until you have submitted the form.
        :::

    1. Set <Hlt>Homepage URL</Hlt> to a URL of your choosing. The value of this
       field is unimportant.

    1. Skip the <Hlt>Identifying and authorizing users</Hlt> and
       <Hlt>Post installation</Hlt> sections of the form. These are not relevant
       to our present goal.

    1. Set <Hlt>Webhook URL</Hlt> to the URL we
       [retrieved earlier](#retrieving-the-receivers-url).

    1. Set <Hlt>Secret</Hlt> to the value previously assigned to the `secret`
       key of the `Secret` referenced by the
       [webhook receiver's configuration](#configuring-the-receiver).
    
    1. In the <Hlt>Permissions</Hlt> section of the form, expand
       <Hlt>Repository Permissions</Hlt>.

    1. Ensure <Hlt>Contents</Hlt> is set to <Hlt>Read-only</Hlt>.

    1. If you would like to receive events when container images or Helm charts
       are pushed to associated GHCR repositories, set <Hlt>Packages</Hlt> to
       <Hlt>Read-only</Hlt>.

        :::note
        You will only receive events from those GHCR repositories explicitly
        associated with your Git repository.

        For more information on this topic, refer to
        [these GitHub docs](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
        :::

    1. In the <Hlt>Subscribe to events</Hlt> section of the form, ensure
       <Hlt>Push</Hlt> is checked.

        If you would like to receive events when container images or Helm charts
        are pushed to associated GHCR repositories, ensure
        <Hlt>Registry package</Hlt> is also checked.

        :::note
        The events available for selection in this section of the form are
        dynamic and dependent on your selections in the <Hlt>Permissions</Hlt>
        section.

        ![Step 9](./img/apps/9.png "Subscribe to Events")

    1. Click <Hlt>Create GitHub App</Hlt>.
```

[shared secret](https://en.wikipedia.org/wiki/Shared_secret) will be used by
GitHub to sign requests. The receiver will use it to authenticate those requests
by verifying their signatures.

:::note
The following command is suggested for generating a complex shared secret:

```shell
openssl rand -base64 48 | tr -d '=+/' | head -c 32
```

:::

```yaml
apiVersion: v1
kind: Secret
metadata:
  name: gh-wh-secret
  namespace: kargo-demo
stringData:
  secret: <your-secret-here>
---
apiVersion: kargo.akuity.io/v1alpha1
kind: ProjectConfig
metadata:
  name: kargo-demo
  namespace: kargo-demo
spec:
  webhookReceivers: 
    - name: gh-wh-receiver
      github:
        secretRef:
          name: gh-wh-secret
```

## Retrieving the Receiver's URL

Kargo will generate a hard-to-guess URL from the configuration. We can obtain 
this URL using the following command:

```shell
  kubectl \
    get projectconfigs \
    kargo-demo \
    -n kargo-demo \
    -o=jsonpath='{.status.webhookReceivers}'
```

## Registering with Github

There are two options whereby GitHub repositories can be configured to send
events to the webhook receiver:

1. Configure webhooks directly for a single repository.

>     The advantage of this approach is that is is comparatively simple, however,

Reviewer: ```suggestion
    The advantage of this approach is that it is comparatively simple, however,
```

    its large disadvantage is that it is tedious and most likely infeasible to
    repeat this configuration for a large number of repositories.

1. Create a [Github App](https://docs.github.com/en/apps).

    The disadvantage of this approach is that it is comparatively complex,
    however, its large advantage is that once created and configured, the App

>     can be easily [installed](https://docs.github.com/en/apps/using-github-apps/installing-your-own-github-app) into any number of GitHub repositories
>     (belonging to the same account that owns the App).

Reviewer: ```suggestion
    can be easily
    [installed](https://docs.github.com/en/apps/using-github-apps/installing-your-own-github-app)
    into any number of GitHub repositories (belonging to the same account that
    owns the App).
```

       key of the `Secret` referenced by the
       [webhook receiver's configuration](#configuring-the-receiver).

    1. Under <Hlt>Which events would you like to trigger this webhook?</Hlt>:
       Leave <Hlt>Just the push event.</Hlt> selected, unless you would
       like to receive events when container images or Helm charts are
       pushed to associated GHCR repositories.

        To receive such events, select
        <Hlt>Let me select individual events.</Hlt>, then ensure
        <Hlt>Pushes</Hlt> and <Hlt>Packages</Hlt> are both checked.

        :::note
        You will only receive events from those GHCR repositories explicitly
        associated with your Git repository.

        For more information on this topic, refer to
        [these GitHub docs](https://docs.github.com/en/packages/learn-github-packages/connecting-a-repository-to-a-package).
        :::

    1. Ensure <Hlt>Active</Hlt> remains checked.

    1. Click <Hlt>Add webhook</Hlt>.

1. Verify connectivity:

    1. From the <Hlt>Webhooks</Hlt> dashboard, select the webhook you registered
       above.
        ![Step 7](./img/webhooks/7.png "Created")

> 

Reviewer: ```suggestion
```

>     1. Select the <Hlt>Recent Deliveries</Hlt> tab.

Reviewer: ```suggestion
    1. Select the <Hlt>Recent Deliveries</Hlt> tab.

```

        :::

    1. In the <Hlt>Subscribe to events</Hlt> section of the form, ensure
       <Hlt>Push</Hlt> is checked.

        If you would like to receive events when container images or Helm charts
        are pushed to associated GHCR repositories, ensure
        <Hlt>Registry package</Hlt> is also checked.

        :::note
        The events available for selection in this section of the form are
        dynamic and dependent on your selections in the <Hlt>Permissions</Hlt>
        section.
        :::

        ![Step 9](./img/apps/9.png "Subscribe to Events")

    1. Under <Hlt>Where can this GitHub App be installed?</Hlt> ensure
       <Hlt>Only this account</Hlt> is selected.

        :::danger
        If you select the other option (<Hlt>Any account</Hlt>), your
        App will be installable into any repository in GitHub, regardless
        of what account owns it. _All_ repositories into which your App
        is installed will send events to the webhook receiver. __You almost
        certainly do not want this!__
        :::

    1. Click <Hlt>Create GitHub App</Hlt>.

> 1. Verify connectivity:

Reviewer: ```suggestion
1. Verify connectivity:

```
